text/0000-dockerfiles.md

```diff
@@ -27,25 +27,65 @@ This RFC proposes that we replace stackpacks with multi-purpose build-time and r
+A builder may specify any number of executable "hooks" in `/cnb/hooks.d/`.
+Hook files in this directory are executed in the context of app directory and must output Dockerfiles that are applied to the build-time and runtime base images before an image is built.
 
-Both Dockerfiles must accept `base_image` and `build_id` args. The `base_image` arg allows the lifecycle to specify the original base image. The `build_id` arg allows the app developer to bust the cache after a certain layer and must be defaulted to `0`.  
+If a hook exits with a non-zero status value, the build fails. If a hook exits with a zero status value and no output, the hook is ignored. Directories and non-executable files are ignored.
+
+Each executable must be in the format `/cnb/hooks.d/<name>.(build.|run.|)<format>`, where build, run, or empty specify the build-time base image, runtime base image, or both bash images, respectively. The only valid format is `Dockerfile`, although support for, e.g. LLB JSON, could be added in the future. 
+
+A runtime Dockerfile is applied to the selected runtime base image after the detection phase.
+A build-time Dockerfile is applied to the build-time base image before the detection phase.
+
+Both Dockerfiles must accept `base_image` and `build_id` args.
+The `base_image` arg allows the lifecycle to specify the original base image.
+The `build_id` arg allows the app developer to bust the cache after a certain layer and must be defaulted to `0`. When the `$build_id` arg is referenced in a `RUN` instruction, all subsequent layerrs will be rebuilt on the next build (as the value will change).
 
 A runtime base image may indicate that it preserves ABI compatibility by adding the label `io.buildpacks.rebasable=true`. In the case of app-specified Dockerfiles, `io.buildpacks.rebasable=false` is set automatically before `run.Dockerfile` is applied and must be explicitly set to `true` if desired. Rebasing an app without this label set to `true` requires passing a new `--force` flag to `pack rebase`.
 
+
+#### Example: App-specified Dockerfile Hook
+
+This example hook would allow an app to provide runtime and build-time base image extensions as "run.Dockerfile" and "build.Dockerfile."
+The app developer can decide whether the extensions are rebasable.
+
+##### `/cnb/hooks.d/app.build.Dockerfile`
+```
+#!/bin/sh
+cat build.Dockerfile
+```
+##### `/cnb/hooks.d/app.run.Dockerfile`
+```
+#!/bin/sh
+cat run.Dockerfile
+```
+
+#### Example: RPM Dockerfile Hook
+
+This example hook would allow a builder to install RPMs for each language runtime.
+
+Note: The Dockerfiles referenced must disable rebasing, and build times will be slower compared to buildpack-provided runtimes.
+
+##### `/cnb/hooks.d/app.Dockerfile`
+```
+#!/bin/sh
+[[ -f Gemfile.lock ]] && cat /cnb/hooks.d/app.Dockerfile.d/Dockerfile-ruby
+[[ -f package.json ]] && cat /cnb/hooks.d/app.Dockerfile.d/Dockerfile-node
+```
+
 ### Platform-specified Dockerfiles
```

Comment:
Can you please describe what is the `platform-` referenced here vs `app-`  or `builder-` ?
Is an application a developer project ? If yes, how will it be possible to copy the files created locally to the `/cnb/hook.d` folder ?

How hooks will be defined part of a buildpacks or platform ?

Reply:
The section just describes how to use Dockerfiles to extend a stack image. It's not related to hooks. Will clarify.

> How hooks will be defined part of a buildpacks or platform ?

Hooks will be added to a builder as part of `pack create-builder`.